service/history/api/replication/syncworkflowstate.go

```diff
@@ -0,0 +1,293 @@
+		}
+	}()
+
+	if err != nil {
+		return nil, err
+	}
+	mu, err := wfCtx.LoadMutableState(ctx, shardContext)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		logger.Info("new run not found", tag.WorkflowNamespaceID(namespaceId.String()), tag.WorkflowID(workflowId), tag.WorkflowRunID(runId))
+		return nil, nil
+	default:
+		return nil, err
+	}
+	versionHistory, err := versionhistory.GetCurrentVersionHistory(mu.GetExecutionInfo().VersionHistories)
+	if err != nil {
+		return nil, err
+	}
+	releaseFunc(nil)
+	releaseFunc = nil
+	newRunEvents, err := getEventsBlob(ctx, shardContext, versionHistory.BranchToken, common.FirstEventID, common.FirstEventID+1)
+	switch err.(type) {
+	case nil:
+	case *serviceerror.NotFound:
+		logger.Info("new run event not found", tag.WorkflowNamespaceID(namespaceId.String()), tag.WorkflowID(workflowId), tag.WorkflowRunID(runId))
+		return nil, nil
+	default:
+		return nil, err
+	}
+	if len(newRunEvents) == 0 {
+		logger.Info("new run event is empty", tag.WorkflowNamespaceID(namespaceId.String()), tag.WorkflowID(workflowId), tag.WorkflowRunID(runId))
+		return nil, nil
+	}
+	return &replicationpb.NewRunInfo{
+		RunId:      runId,
+		EventBatch: newRunEvents[0],
+	}, nil
+}
+
+func getMutation(mutableState workflow.MutableState, versionedTransition *persistencepb.VersionedTransition) (*persistencepb.WorkflowMutableStateMutation, error) {
+	rootNode := mutableState.HSM()
+	updatedStateMachine, err := getUpdatedSubStateMachine(rootNode, versionedTransition)
+	if err != nil {
+		return nil, err
+	}
+	tombstoneBatch := mutableState.GetExecutionInfo().SubStateMachineTombstoneBatches
+	var tombstones []*persistencepb.StateMachineTombstoneBatch
+	for i, tombstone := range tombstoneBatch {
+		if workflow.CompareVersionedTransition(tombstone.VersionedTransition, versionedTransition) >= 0 {
+			tombstones = tombstoneBatch[i:]
+			break
+		}
+	}
+	executionInfo := mutableState.GetExecutionInfo()
+	executionInfo.UpdateInfos = nil
+	executionInfo.SubStateMachinesByType = nil
+	return &persistencepb.WorkflowMutableStateMutation{
+		UpdatedActivityInfos:            getUpdatedInfo(mutableState.GetPendingActivityInfos(), versionedTransition),
+		UpdatedTimerInfos:               getUpdatedInfo(mutableState.GetPendingTimerInfos(), versionedTransition),
+		UpdatedChildExecutionInfos:      getUpdatedInfo(mutableState.GetPendingChildExecutionInfos(), versionedTransition),
+		UpdatedRequestCancelInfos:       getUpdatedInfo(mutableState.GetPendingRequestCancelExternalInfos(), versionedTransition),
+		UpdatedSignalInfos:              getUpdatedInfo(mutableState.GetPendingSignalExternalInfos(), versionedTransition),
+		UpdatedUpdateInfos:              getUpdatedInfo(mutableState.GetExecutionInfo().UpdateInfos, versionedTransition),
+		UpdatedSubStateMachines:         updatedStateMachine,
+		SubStateMachineTombstoneBatches: tombstones,
+		SignalRequestedIds:              mutableState.GetPendingSignalRequestedIds(),
```

Comment:
only include this when the SignalRequestedIdLastUpdateVersionedTransition is new enough.